core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -375,6 +377,208 @@ int spmc_sp_add_share(struct ffa_rxtx *rxtx,
+		  sizeof(struct ffa_mem_access) +
+		  sizeof(struct ffa_mem_region);
+
+	if (tx_len < 0)
+		return FFA_NO_MEMORY;
+
+	SLIST_FOREACH(reg, &smem->regions, link) {
+		tx_len -= sizeof(struct ffa_address_range);
+		if (tx_len < 0)
+			return FFA_NO_MEMORY;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void create_retrieve_response(void *dst_buffer,
+				     struct sp_mem_receiver *receiver,
+				     struct sp_mem *smem, struct sp_session *s)
+{
+	size_t off = 0;
+	struct ffa_mem_region *dst_region =  NULL;
+	struct ffa_mem_transaction *d_ds = dst_buffer;
+	struct ffa_address_range *addr_dst = NULL;
+	struct sp_mem_map_region *reg = NULL;
+
+	/*
+	 * We respond with a FFA_MEM_RETRIEVE_RESP which defines the
+	 * following data in the rx buffer of the SP.
+	 * struct mem_transaction_descr
+	 * struct mem_access_descr //1 Element
+	 * struct mem_region_descr
+	 */
+	/* Copy the mem_transaction_descr */
+	d_ds->sender_id = receiver->smem->sender_id;
+	d_ds->mem_reg_attr = receiver->smem->mem_reg_attr;
+	d_ds->flags = receiver->smem->flags;
+	d_ds->tag = receiver->smem->tag;
+
+	off = sizeof(struct ffa_mem_transaction) +
+	      sizeof(struct ffa_mem_access);
+
+	d_ds->mem_access_count = 1;
+
+	/* Copy the mem_accsess_descr */
+	d_ds->mem_access_array[0].region_offs = off;
+	memcpy(&d_ds->mem_access_array[0].access_perm,
+	       &receiver->perm, sizeof(struct ffa_mem_access_perm));
+
+	/* Copy the mem_region_descr */
+	dst_region = (struct ffa_mem_region *)((vaddr_t)d_ds + off);
+
+	dst_region->address_range_count = 0;
+
+	addr_dst = &dst_region->address_range_array[0];
+
+	dst_region->total_page_count = 0;
+
+	SLIST_FOREACH(reg, &smem->regions, link) {
+		uint32_t offset = reg->page_offset;
+		struct sp_ctx *ctx = to_sp_ctx(s->ts_sess.ctx);
+
+		addr_dst->address = (uint64_t)sp_mem_get_va(&ctx->uctx,
+							    offset,
+							    reg->mobj);
+		addr_dst->page_count = reg->page_count;
+		dst_region->address_range_count++;
+
+		dst_region->total_page_count += addr_dst->page_count;
+	}
+}
+
+static void ffa_mem_retrieve(struct thread_smc_args *args,
+			     struct sp_session *caller_sp,
+			     struct ffa_rxtx *rxtx)
+{
+	TEE_Result ret = FFA_OK;
+	size_t tx_len = 0;
+	struct ffa_mem_transaction *retr_dsc = NULL;
+	struct ffa_mem_region *mem_region = NULL;
+	uint64_t va = 0;
+	struct sp_mem *smem = NULL;
+	struct sp_mem_receiver *receiver = NULL;
+	uint32_t exceptions = 0;
+
+	if (!check_rxtx(rxtx) || !rxtx->tx_is_mine) {
+		ret = FFA_DENIED;
+		goto out;
+	}
+
+	tx_len = rxtx->size;
+	retr_dsc = rxtx->rx;
+
+	smem = sp_mem_get(retr_dsc->global_handle);
+
+	if (!smem) {
+		DMSG("Incorrect handle");
+		ret = FFA_DENIED;
+		goto out;
+	}
+
+	receiver = sp_mem_get_receiver(caller_sp->endpoint_id, smem);
+
+	if (check_retrieve_request(receiver, retr_dsc, smem, tx_len) !=
+	    TEE_SUCCESS) {
+		ret = FFA_INVALID_PARAMETERS;
+		goto out;
+	}
+
+	exceptions = cpu_spin_lock_xsave(&mem_ref_lock);
+	receiver->ref_count++;
```

Comment:
This is a `uint8_t`, any risk that it can wrap?

Reply:
Added a check